teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+### Per feature flag for _new_ api-driven features
+Introduce per-feature flags for each **new** API driven feature. Each feature will have its own flag, instead of using the group API driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing beahviour.
+
+The flag will also include the maintainer information and stability level for the feature as the source of truth. New behavioural features will also have a new per-feature flag to either enable or disable the feature. This will allow behavioural features that has values leading to behaviors in different stability levels to be turned on or off instead of depending on the stability level of features that are enabled.
+
+For example:
+
+```yaml
+apiVersion: `v1`
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  # alpha: v0.53
+  #
+  # Pipeline in pipeline has ... functionalities.
+  # It is disabled by default.
+  enable-pipeline-in-pipeline: "false"
+  # alpha: v0.53
+  # 
+  # Trusted artifacts has ... functionalities.
+  # It is disabled by default. Set it to true to enable this feature.
+  enable-trusted-artifacts: "true"
+```
+
+See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.
+
+All **new** features can only be enabled via per-feature flags. When they first get introduced as alpha, they will be disabled by default. When new features get promoted to stable, they will be enabled by default according to the following table:
+
+| Feature stability level | Default  |
+| ----------------------- | -------- |
+| Stable                  | Enabled; Cannot be disabled  |
+| Beta                    | Disabled |
+| Alpha                   | Disabled |
+
+Note that for per feature flag that has stabilized, they cannot be disabled and we will remove the feature flag after 3 releases after it has become stable. Notification will be given to users via release notes.
+
+The behaviour of existing `enable-api-fields` flag with per feature flag:
+- Any current beta features can be enabled with `enable-api-fields` set to “beta” or “alpha”.
+- Any current alpha features can be enabled with `enable-api-fields` set to “alpha”. When current alpha features are promoted to beta, they can be enabled with `enable-api-fields` set to `beta` or `alpha`.
+- It will not be possible to enable existing features using per-feature flags.
```

Comment:
```suggestion
- The existing features are not adopting this new mechanism of having one flag per feature. It will not be possible to enable existing features using per-feature flags.
```

Reply:
Thank you for the suggestion!